File: java/src/main/structural/proxy/Proxy.java
package structural.proxy;

// Proxy: Provides controlled access to the RealSubject
public class Proxy implements Subject {
    private RealSubject realSubject;

    @Override
    public void request() {
        System.out.println("Proxy: Checking access before delegating to RealSubject.");
        
        if (realSubject == null) {
            System.out.println("Proxy: Initializing RealSubject for the first time.");
            realSubject = new RealSubject();
        } else {
            System.out.println("Proxy: Using existing RealSubject instance.");
        }

        realSubject.request();
        System.out.println("Proxy: Logging the request.");
    }
}

// Example Usage:
// public class Main {
//     public static void main(String[] args) {
//         Subject proxy = new Proxy();
//
//         // Client interacts with the Proxy, not the RealSubject
//         proxy.request();
//     }
// }
//
// Common Mistakes to Avoid:
//
// 1. Overcomplicating the Proxy:
//    - Mistake: Adding excessive logic to the Proxy, making it harder to maintain.
//    - Fix: Keep the Proxy focused on controlling access and delegating requests.
//
// 2. Ignoring Lazy Initialization:
//    - Mistake: Always initializing the RealSubject, defeating the purpose of the Proxy.
//    - Fix: Use lazy initialization to create the RealSubject only when needed.
//
// 3. Breaking Encapsulation:
//    - Mistake: Allowing direct access to the RealSubject from outside the Proxy.
//    - Fix: Ensure all access goes through the Proxy.
//
// 4. Inefficient Access Control:
//    - Mistake: Using inefficient checks or redundant logic in the Proxy.
//    - Fix: Optimize access control mechanisms to avoid performance overhead.

File: python/src/structural/proxy/proxy.py
from src.structural.proxy.subject import Subject
from src.structural.proxy.real_subject import RealSubject


class Proxy(Subject):
    def __init__(self):
        self._real_subject = None

    def request(self) -> str:
        if self._real_subject is None:
            print("Proxy: Creating RealSubject...")
            self._real_subject = RealSubject()
        print("Proxy: Delegating to RealSubject.")
        return self._real_subject.request()

# Example Usage:
# if __name__ == "__main__":
#     proxy = Proxy()
#     print(proxy.request())  # Lazy-loads RealSubject and delegates
#     print(proxy.request())  # Reuses the same RealSubject

# Common Mistakes to Avoid:
#
# 1. Not implementing the same interface:
#    - Mistake: Proxy and RealSubject don’t share the Subject base class.
#    - Fix: Always make both implement the same interface for transparency.
#
# 2. Forgetting lazy initialization:
#    - Mistake: Creating the RealSubject eagerly when it’s expensive.
#    - Fix: Delay creation until absolutely necessary (lazy loading).
#
# 3. Violating the open-closed principle:
#    - Mistake: Tightly coupling Proxy logic with RealSubject internals.
#    - Fix: Proxy should treat RealSubject as a black box.
#
# 4. Using Proxy where not needed:
#    - Mistake: Adding complexity for simple delegation or wrapper.
#    - Fix: Only use Proxy when you need additional control over access.
